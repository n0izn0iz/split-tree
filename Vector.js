import Space from "./Space";

export default class Vector extends Array {
  constructor(...values) {
    const space = values[0];
    if (space instanceof Space) super(...space.originVector);
    else {
      if (values.length === 1) {
        super(1);
        this[0] = values[0];
      } else super(...values);
    }
    this.__proto__ = Vector.prototype;
    if (this.length === 0) throw new Error("Created 0 dimensions Vector");
  }
  toString() {
    return `${this.length}DVector: { ${[...this]
      .map((value, index) => `${["x", "y", "z", "t"][index]}: ${value}`)
      .join(", ")} }`;
  }
  mapToVector(...args) {
    return new Vector(...this.map(...args));
  }
  plusScalar(scalar) {
    return this.mapToVector(value => value + scalar);
  }
  minusScalar(scalar) {
    return this.mapToVector(value => value - scalar);
  }
  plus(other) {
    if (Array.isArray(other)) other = new Vector(...other);
    const biggestVector = this.length > other.length ? this : other;
    const smallestVector = biggestVector === other ? this : other;
    return biggestVector.mapToVector(
      (value, index) =>
        value + (index < smallestVector.length ? smallestVector[index] : 0)
    );
  }
  scale(scalar) {
    return this.mapToVector(value => value * scalar);
  }
}
