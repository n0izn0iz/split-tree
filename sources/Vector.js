// @flow
export default class Vector extends Array<number> {
  constructor(...values: Array<number>) {
    if (values.length === 1) {
      super();
      this[0] = values[0];
    } else {
      super(...values);
    }
    // $FlowFixMe
    this.__proto__ = Vector.prototype;
    //if (this.length === 0) throw new Error("Created 0 dimensions Vector");
  }
  toString() {
    return `${this.length}DVector: { ${[...this]
      .map((value, index) => `${["x", "y", "z", "t"][index]}: ${value}`)
      .join(", ")} }`;
  }

  get array(): Array<number> {
    return [...this];
  }

  get time(): number {
    return this[this.length - 1];
  }

  get x(): number {
    return this[0];
  }
  get y(): number {
    return this[1];
  }
  set y(newValue: number) {
    this[1] = newValue;
  }
  mapToVector(...args: Array<any>): Vector {
    return new Vector(...this.map(...args));
  }
  multiply(other: Vector) {
    return this.mapToVector((value, index) => value * other[index]);
  }
  plusScalar(scalar: number) {
    return this.mapToVector(value => value + scalar);
  }
  minusScalar(scalar: number) {
    return this.mapToVector(value => value - scalar);
  }
  reverseVector() {
    return new Vector(...super.reverse());
  }
  plus(other: Vector) {
    if (Array.isArray(other)) other = new Vector(...other);
    const biggestVector = this.length > other.length ? this : other;
    const smallestVector = biggestVector === other ? this : other;
    return biggestVector.mapToVector(
      (value, index) =>
        value + (index < smallestVector.length ? smallestVector[index] : 0)
    );
  }
  minus(other: Vector) {
    return this.plus(other.scale(-1));
  }
  scale(scalar: number) {
    return this.mapToVector(value => value * scalar);
  }
  normalize() {
    const length = Math.sqrt(
      this.reduce((lengthSquared, value) => {
        return lengthSquared + value * value;
      }, 0)
    );
    return this.mapToVector(value => value / length);
  }
  inequal(other: Vector) {
    return this.some((value, index) => value !== other[index]);
  }
  equal(other: Vector) {
    return !this.inequal(other);
  }
  lesser(other: Vector) {
    return !this.some((value, index) => value >= other[index]);
  }
  lesserOrEqual(other: Vector) {
    if (this.equal(other)) return true;
    return this.lesser(other);
  }
  greater(other: Vector) {
    return !this.lesserOrEqual(other);
  }

  greaterOrEqual(other: Vector) {
    if (this.equal(other)) return true;
    return this.greater(other);
  }
}
