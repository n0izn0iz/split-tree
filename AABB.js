// @flow
import Vector from "./Vector";
// import Direction from "td-direction";

const minOrMax = mm => (mm ? "min" : "max");

export default class AABB {
  min: Vector;
  max: Vector;
  constructor(
    position: Array<number>,
    options: {
      size?: number,
      clone?: ?AABB,
      min?: ?Array<number>,
      max?: ?Array<number>
    }
  ) {
    const { size = 1, min = null, max = null, clone = null } = options;
    if (clone !== null) {
      this.min = clone.min;
      this.max = clone.max;
    } else if (!(min === null && max === null)) {
      if (min === null || max === null)
        throw new Error(
          `Constructing aABB with ${minOrMax(min)} but no ${minOrMax(max)}`
        );
      this.min = new Vector(...min);
      this.max = new Vector(...max);
    } else if (!position) {
      throw new Error("Constructing aABB without position");
    } else {
      const halfSize = size / 2;
      const positionVector = new Vector(...position);
      this.min = positionVector.minusScalar(halfSize);
      this.max = positionVector.plusScalar(halfSize);
    }
  }

  toString() {
    return `${this
      .dimensionsCount}DAABB: { min: ${this.min.toString()}, max: ${this.max.toString()} }`;
  }

  /*quadrant(directionIndex: number) {
    const position = this.position;

    switch (directionIndex) {
      case Direction.topLeftIndex:
        const left = vec2.fromValues(this.min[0], position[1]);
        const top = vec2.fromValues(position[0], this.max[1]);
        return new AABB({ min: left, max: top });

      case Direction.maxIndex:
        return new AABB({ min: position, max: this.max });

      case Direction.bottomRightIndex:
        const bottom = vec2.fromValues(position[0], this.min[1]);
        const right = vec2.fromValues(this.max[0], position[1]);
        return new AABB({ min: bottom, max: right });

      case Direction.minIndex:
        return new AABB({ min: this.min, max: position });
    }
  }*/

  // $FlowFixMe
  set position(position: Array<number>) {
    const aABB = new AABB(position, { size: this.size });
    this.min = aABB.min;
    this.max = aABB.max;
  }

  // $FlowFixMe
  get position(): Array<number> {
    return this.min.plusScalar(this.halfSize);
  }

  // $FlowFixMe
  get halfSize(): number {
    return this.size / 2;
  }

  // $FlowFixMe
  get size(): number {
    return this.max[0] - this.min[0];
  }

  // $FlowFixMe
  collideAABB(other: AABB) {
    return this.points.some(point => other.containsPoint(point));
  }

  static __points(currentPoint, currentDimension, size) {
    const nextPoint = currentPoint.mapToVector(
      (value, index) => (index === currentDimension - 1 ? value + size : value)
    );
    if (currentDimension === 1)
      return [currentPoint.reverse(), nextPoint.reverse()];
    else
      return [
        ...AABB.__points(currentPoint, currentDimension - 1, size),
        ...AABB.__points(nextPoint, currentDimension - 1, size)
      ];
  }

  // $FlowFixMe
  get points(): Array<Array<number>> {
    return AABB.__points(this.min, this.dimensionsCount, this.size);
  }

  // $FlowFixMe
  get dimensionsCount(): number {
    return this.min.length;
  }

  containsPoint(point: Array<number>) {
    if (point.length !== this.dimensionsCount)
      throw "Dimensions mismatch: Called containsPoint on " +
        this.toString() +
        " with " +
        point.toString();
    return !point.some((value, index) => {
      return value < this.min[index] || value > this.max[index];
    });
  }

  containsAABB(other: AABB) {
    return (
      other.min[0] >= this.min[0] &&
      other.min[1] >= this.min[1] &&
      other.max[0] <= this.max[0] &&
      other.max[1] <= this.max[1]
    );
  }
}
