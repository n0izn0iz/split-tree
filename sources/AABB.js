//@flow
import Vector from "./Vector";
// import Direction from "td-direction";

const minOrMax = mm => (mm ? "min" : "max");

const __points = (currentPoint, currentDimension, extents): Array<Vector> => {
  const nextPoint = currentPoint.mapToVector(
    (value, index) =>
      index === currentDimension - 1 ? value + extents[index] : value
  );
  if (currentDimension === 1) return [currentPoint, nextPoint];
  else
    return [
      ...__points(currentPoint, currentDimension - 1, extents),
      ...__points(nextPoint, currentDimension - 1, extents)
    ];
};

export default class AABB {
  min: Vector;
  max: Vector;
  __pointsCache: ?Array<Vector>;
  __partitionsCache: ?Array<AABB>;
  constructor(position: Vector, extents: Vector) {
    const halfExtents = extents.scale(0.5);
    this.min = position.minus(halfExtents);
    this.max = position.plus(halfExtents);
  }

  toString() {
    return `${this
      .dimensionsCount}DAABB: { min: ${this.min.toString()}, max: ${this.max.toString()} }`;
  }

  set position(position: Vector) {
    const aABB = new AABB(position, this.extents);
    this.min = aABB.min;
    this.max = aABB.max;
    this.__pointsCache = null;
    this.__partitionsCache = null;
  }

  get position(): Vector {
    return this.min.plus(this.halfExtents);
  }

  get halfExtents(): Vector {
    return this.extents.scale(0.5);
  }

  get extents(): Vector {
    return this.max.minus(this.min);
  }

  collideAABB(other: AABB) {
    return other.points.some(point => this.containsPoint(point));
  }

  get partitions(): Array<AABB> {
    if (!this.__partitionsCache) {
      const position = this.position;
      this.__partitionsCache = this.localPoints.map(localPoint => {
        return new AABB(localPoint.scale(0.5).plus(position), this.halfExtents);
      });
    }
    return this.__partitionsCache;
  }

  partition(index: number): AABB {
    /*if (index >= this.partitionsCount)
      throw new Error(
        `Dimensions mismatch: Called partition(${index}) on ${this
          .dimensionsCount}D AABB, maximum index: ${this.partitionsCount - 1}`
      );*/
    const partition = this.partitions[index];
    return new AABB(partition.position, partition.extents);
  }

  get partitionsCount(): number {
    return Math.pow(this.dimensionsCount, 2);
  }

  get points(): Array<Vector> {
    if (!this.__pointsCache)
      this.__pointsCache = __points(
        this.min,
        this.dimensionsCount,
        this.extents
      );
    return this.__pointsCache;
  }

  get localMin(): Vector {
    const { extents } = this;
    return this.extents.mapToVector(value => -(value / 2));
  }

  get localPoints(): Array<Vector> {
    return __points(this.localMin, this.dimensionsCount, this.extents);
  }

  get dimensionsCount(): number {
    return this.min.length;
  }

  containsPoint(point: Vector) {
    if (point.length !== this.dimensionsCount)
      throw "Dimensions mismatch: Called containsPoint on " +
        this.toString() +
        " with " +
        point.toString();
    return point.every(
      (value, index) => this.min[index] <= value && value <= this.max[index]
    );
  }

  containsAABB(other: AABB) {
    if (this.dimensionsCount !== other.dimensionsCount)
      throw "Dimensions mismatch: Called containsAABB on " +
        this.toString() +
        " with " +
        other.toString();
    return this.containsPoint(other.min) && this.containsPoint(other.max);
  }
}
