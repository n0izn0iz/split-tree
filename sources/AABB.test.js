import AABB from "./AABB";
import Vector from "./Vector";

test("constructor throws without arguments", () => {
  expect(() => {
    new AABB();
  }).toThrow();
});

test("constructor matches snapshots with various numbers", () => {
  expect(new AABB(new Vector(42), new Vector(1))).toMatchSnapshot();
  expect(new AABB(new Vector(1, 2), new Vector(1, 1))).toMatchSnapshot();
  expect(
    new AABB(new Vector(78, 45, 12), new Vector(1, 1, 1))
  ).toMatchSnapshot();
});

test("partitions matches snapshots with various numbers", () => {
  expect(new AABB(new Vector(42), { size: 1 }).partitions).toMatchSnapshot();
  expect(new AABB(new Vector(1, 2), { size: 1 }).partitions).toMatchSnapshot();
  expect(
    new AABB(new Vector(78, 45, 12), { size: 1 }).partitions
  ).toMatchSnapshot();
});

test("points matches snapshots with various numbers", () => {
  expect(new AABB(new Vector(42), { size: 1 }).points).toMatchSnapshot();
  expect(new AABB(new Vector(1, 2), { size: 1 }).points).toMatchSnapshot();
  expect(
    new AABB(new Vector(78, 45, 12), { size: 1 }).points
  ).toMatchSnapshot();
});
