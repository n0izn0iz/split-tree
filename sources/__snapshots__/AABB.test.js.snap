// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor matches snapshots with various numbers 1`] = `
AABB {
  "max": Vector [
    42.5,
  ],
  "min": Vector [
    41.5,
  ],
}
`;

exports[`constructor matches snapshots with various numbers 2`] = `
AABB {
  "max": Vector [
    1.5,
    2.5,
  ],
  "min": Vector [
    0.5,
    1.5,
  ],
}
`;

exports[`constructor matches snapshots with various numbers 3`] = `
AABB {
  "max": Vector [
    78.5,
    45.5,
    12.5,
  ],
  "min": Vector [
    77.5,
    44.5,
    11.5,
  ],
}
`;

exports[`partitions matches snapshots with various numbers 1`] = `
Array [
  AABB {
    "max": Vector [
      42,
    ],
    "min": Vector [
      41.5,
    ],
  },
  AABB {
    "max": Vector [
      42.5,
    ],
    "min": Vector [
      42,
    ],
  },
]
`;

exports[`partitions matches snapshots with various numbers 2`] = `
Array [
  AABB {
    "max": Vector [
      1,
      2,
    ],
    "min": Vector [
      0.5,
      1.5,
    ],
  },
  AABB {
    "max": Vector [
      1.5,
      2,
    ],
    "min": Vector [
      1,
      1.5,
    ],
  },
  AABB {
    "max": Vector [
      1,
      2.5,
    ],
    "min": Vector [
      0.5,
      2,
    ],
  },
  AABB {
    "max": Vector [
      1.5,
      2.5,
    ],
    "min": Vector [
      1,
      2,
    ],
  },
]
`;

exports[`partitions matches snapshots with various numbers 3`] = `
Array [
  AABB {
    "max": Vector [
      78,
      45,
      12,
    ],
    "min": Vector [
      77.5,
      44.5,
      11.5,
    ],
  },
  AABB {
    "max": Vector [
      78.5,
      45,
      12,
    ],
    "min": Vector [
      78,
      44.5,
      11.5,
    ],
  },
  AABB {
    "max": Vector [
      78,
      45.5,
      12,
    ],
    "min": Vector [
      77.5,
      45,
      11.5,
    ],
  },
  AABB {
    "max": Vector [
      78.5,
      45.5,
      12,
    ],
    "min": Vector [
      78,
      45,
      11.5,
    ],
  },
  AABB {
    "max": Vector [
      78,
      45,
      12.5,
    ],
    "min": Vector [
      77.5,
      44.5,
      12,
    ],
  },
  AABB {
    "max": Vector [
      78.5,
      45,
      12.5,
    ],
    "min": Vector [
      78,
      44.5,
      12,
    ],
  },
  AABB {
    "max": Vector [
      78,
      45.5,
      12.5,
    ],
    "min": Vector [
      77.5,
      45,
      12,
    ],
  },
  AABB {
    "max": Vector [
      78.5,
      45.5,
      12.5,
    ],
    "min": Vector [
      78,
      45,
      12,
    ],
  },
]
`;

exports[`points matches snapshots with various numbers 1`] = `
Array [
  Vector [
    41.5,
  ],
  Vector [
    42.5,
  ],
]
`;

exports[`points matches snapshots with various numbers 2`] = `
Array [
  Vector [
    0.5,
    1.5,
  ],
  Vector [
    1.5,
    1.5,
  ],
  Vector [
    0.5,
    2.5,
  ],
  Vector [
    1.5,
    2.5,
  ],
]
`;

exports[`points matches snapshots with various numbers 3`] = `
Array [
  Vector [
    77.5,
    44.5,
    11.5,
  ],
  Vector [
    78.5,
    44.5,
    11.5,
  ],
  Vector [
    77.5,
    45.5,
    11.5,
  ],
  Vector [
    78.5,
    45.5,
    11.5,
  ],
  Vector [
    77.5,
    44.5,
    12.5,
  ],
  Vector [
    78.5,
    44.5,
    12.5,
  ],
  Vector [
    77.5,
    45.5,
    12.5,
  ],
  Vector [
    78.5,
    45.5,
    12.5,
  ],
]
`;
